File: Solutions/206-Reverse-Linked-List.cs

// recursive with two parameters
public class Solution {
    public ListNode ReverseList(ListNode head, ListNode? last = null) {
        if(head is null) return last;

        var next = head.next;
        head.next = last;
        return ReverseList(next, head);
    }
}

// response recursive
public class Solution {
    public ListNode ReverseList(ListNode head) {
        if(head?.next is null) return head;

        var reversedSublist = ReverseList(head.next);
        
        head.next.next = head;
        head.next = null;

        return reversedSublist;    
    }
}

File: Solutions/206-Reverse-Linked-List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head

        prevNode: Optional[ListNode] = None
        nextNode: Optional[ListNode] = None
        currentNode: Optional[ListNode] = head

        while(currentNode.next):
            nextNode = currentNode.next #2 3
            currentNode.next = nextNode.next #3 4
            nextNode.next = head #1 2
            head = nextNode # 2 3

        return head
